caffe2/python/pybind_state.cc

```diff
@@ -422,7 +422,13 @@ void addObjectMethods(py::module& m) {
       .def("_wrap_tensor_impl", [](Blob* blob, void* ptr) {
         auto p = c10::intrusive_ptr<c10::TensorImpl, at::UndefinedTensorImpl>::
             unsafe_reclaim_from_nonowning(static_cast<c10::TensorImpl*>(ptr));
+        // TODO: In the near future, a PyTorch tensor without AutogradMeta will be
+        // a valid tensor. At that point, we will only accept no-autograd-meta
+        // tensor into Caffe2 workspace, and don't need to perform shallow-copying
+        // here anymore.
+        p = p->shallow_copy_and_detach();
```

Comment:
@gchanan I think we should do shallow-copying here instead of directly detaching AutogradMeta from `p`, because the latter will break the invariant that a PyTorch tensor always has AutogradMeta and turn `p` into an invalid PyTorch tensor. Consider an example:
```python
z = torch.ones((4,), dtype=torch.int64)
workspace.FeedBlob('bar', z)  # In this step, we remove `z`'s AutogradMeta
del z  # Segfault
```
The last line `del z` will segfault at `THPVariable_clear()`, because it tries to access `z`'s AutogradMeta but it doesn't exist anymore. To maintain the invariant that a PyTorch tensor always has AutogradMeta, we should not directly detach AutogradMeta here, and should instead do shallow-copying.

Reply:
ya, I thought there was an API to read the PT tensor back out from the workspace at the boundary, but that's not applicable here.